packages/wallet/src/KeyManagement/util/mapHardwareSigningData.ts

```diff
@@ -86,6 +128,320 @@ const bytesToIp = (bytes?: Uint8Array) => {
   return null;
 };
 
+const prepareTrezorInputs = (
+  inputs: CSL.TransactionInputs,
+  inputAddressResolver: ResolveInputAddress,
+  knownAddresses: GroupedAddress[]
+): CardanoInput[] => {
+  const trezorInputs = [];
+  for (let i = 0; i < inputs.len(); i++) {
+    const input = inputs.get(i);
+    const coreInput = cslToCore.txIn(input);
+    const paymentAddress = inputAddressResolver(coreInput);
+
+    let trezorInput = {
+      prev_hash: Buffer.from(input.transaction_id().to_bytes()).toString('hex'),
+      prev_index: input.index()
+    } as CardanoInput;
+
+    let paymentKeyPath = null;
+    if (paymentAddress) {
+      const knownAddress = knownAddresses.find(({ address }) => address === paymentAddress);
+      if (knownAddress) {
+        paymentKeyPath = [
+          HARDENED + 1852,
+          HARDENED + 1815,
+          HARDENED + knownAddress.accountIndex,
+          knownAddress.type,
+          knownAddress.index
+        ];
+        trezorInput = {
+          ...trezorInput,
+          path: paymentKeyPath
+        };
+      }
+    }
+    trezorInputs.push(trezorInput);
+  }
+  return trezorInputs;
+};
+
+const prepareTrezorOutputs = (
+  outputs: CSL.TransactionOutputs,
+  accountIndex: number,
+  knownAddresses: GroupedAddress[]
+): CardanoOutput[] => {
+  const trezorOutputs = [];
+  for (let i = 0; i < outputs.len(); i++) {
+    const output = outputs.get(i);
+    const multiAsset = output.amount().multiasset();
+    const tokenBundle = [];
+    if (multiAsset) {
+      for (let j = 0; j < multiAsset.keys().len(); j++) {
+        const policy = multiAsset.keys().get(j);
+        const assets = multiAsset.get(policy);
+        const tokens = [];
+        if (assets) {
+          for (let k = 0; k < assets.keys().len(); k++) {
+            const assetName = assets.keys().get(k);
+            const amount = assets.get(assetName);
+            if (assetName && amount) {
+              tokens.push({
+                amount: amount.to_str(),
+                assetNameBytes: Buffer.from(assetName.name()).toString('hex')
+              });
+            }
+          }
+        }
+        sortTrezorTokensCanonically(tokens);
+        tokenBundle.push({
+          policyId: Buffer.from(policy.to_bytes()).toString('hex'),
+          tokenAmounts: tokens
+        });
+      }
+    }
+    const outputAddress = Buffer.from(output.address().to_bytes());
+    const isDeviceOwned = knownAddresses.some(
+      ({ address }) => address.toString() === utils.bech32_encodeAddress(outputAddress)
+    );
+    const destination = isDeviceOwned
+      ? {
+          addressParameters: {
+            addressType: CardanoAddressType.BASE,
+            path: `m/1852'/1815'/${accountIndex}'/0/0`,
```

Comment:
I think `role` and `index` should come from `knownAddresses` instead of always being 0/0 here